Parallelize bi-LSTM implementation

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21582 Parallelize bi-LSTM implementation**

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

Differential Revision: [D15737066](https://our.internmc.facebook.com/intern/diff/D15737066)

Commit:
Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

ilia-cher commented:
testing with nmt benchmark

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

dzhulgakov commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -365,23 +371,39 @@ struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<
+    using unstacked_output_type = typename FullLayer<dir_hidden_type, cell_params>::unstacked_output_type;
+    unstacked_output_type fw_result;
+    at::Tensor fw_output;
+    at::intraop_launch([&]() {
```
can it (optionally) return some Future-like abstraction? it's annoying to reimplement the same pattern in many places. How about ivalue::Future?

though in this case it can be just regular std::promise<void>, right?

ilia-cher replied:
There's an annoying issue in ivalue::Future that we need to fix first, it forces us to use std::shared_ptr to pass a future inside lambda and I don't want to add this overhead atm

the issue happens when the future is defined locally somewhere before lambda, then passed into lambda, then we schedule lambda, then we wait on future after scheduling lambda.

At some moment 'markCompleted' is going to be called from inside the lambda. Our current 'wait' is implemented as a callback.
While future still executes markCompleted it will unblock waiter that will finish the work and return and destroy the future.

My plan was to address this separately and provide a more convenient API, but that's not blocking atm, we can simply use condition_variable here

ilia-cher replied:
actually, let me check std::future/std::promise - maybe it will be possible to use it from intraop_launch

ilia-cher replied:
I think we could also use std future in ParallelNative.h too

ilia-cher replied:
@dzhulgakov sent https://github.com/pytorch/pytorch/pull/21764

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

cpuhrsch commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -336,12 +337,13 @@ struct FullLayer : Layer<Tensor, hidden_type, cell_params> {
-      hidden = cell_(step_inputs[i], hidden, params);
-      step_outputs.push_back(hidden_as_output(hidden));
+    for (size_t idx = 0; idx < step_inputs.size(); idx++) {
+      size_t input_idx = reverse ? step_inputs.size() - 1 - idx : idx;
```
Does this actually help in benchmarks?

cpuhrsch replied:
Mostly just curious

ilia-cher replied:
it seems it's about the same speed or slightly faster based on benchmark

cpuhrsch commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -365,23 +367,42 @@ struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<
+#else
+    unstacked_output_type fw_result;
+    at::Tensor fw_output;
+    auto fut = at::intraop_launch_future([&]() {
```
Is this actually a good idea for something like openmp? Won't it execute two OpenMP operators concurrently which might oversubscribe the system?

ilia-cher replied:
in case of openmp this will be inline, oversubscribing depends on the execution environment - in some cases we can benefit form extra parallelizm, but not always, depending on whether there're other inter-op tasks, how many parallel requests we handle, are there other processes running, machine capability, this needs to be benchmarked each time

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head

Commit:
Update on "Parallelize bi-LSTM implementation"

Parallelize bi-LSTM implementation

Summary:
Using at::intraop_launch to run fwd and rev passes simultaneously

gh-metadata: pytorch pytorch 21582 gh/ilia-cher/27/head
